src/Features/CSharp/Portable/CodeRefactorings/NamedArguments/CSharpAddNamedArgumentsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.NamedArguments;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.NamedArguments
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpAddNamedArgumentsCodeRefactoringProvider)), Shared]
+    internal class CSharpAddNamedArgumentsCodeRefactoringProvider : AbstractAddNamedArgumentsCodeRefactoringProvider<ArgumentSyntax>
+    {
+        protected override IEnumerable<ArgumentSyntax> GetNamedArguments(IList<ArgumentSyntax> arguments, ImmutableArray<IParameterSymbol> parameters)
+        {
+            if (arguments == null)
+            {
+                throw new ArgumentNullException(nameof(arguments));
+            }
+
+            for (int index = 0; index < arguments.Count; ++index)
+            {
+                var parameter = parameters[index];
+                var argument = arguments[index];
+
+                if (argument.NameColon != null)
+                {
+                    yield return argument;
+                }
+                else if (parameter.IsParams && arguments.Count - index > 1)
+                {
+                    var expressions = arguments.Skip(index).Select(arg => arg.Expression);
```

Comment:
> Including 'params' arguments

How to not include it? `params` comes last so do named arguments, the only option is to reorder arguments, that is what you are suggesting?

> Only for literal arguments
 
Same here, because they might be in any arbitrary order. Would you please tell me what exact rules do you have in mind? thanks.